### PR TITLE
Ruby 3.2 compatibility ✅

### PR DESCRIPTION
**Pull request by thiagofm**
For the invoker classes, we started to rely on `.parameters` to know if there are keyword arguments. It's a clearer approach than relying on arity. I couldn't make it work by just relying on arity.

I've also improved test coverage a bit, those new test cases were all failing because of the Ruby 3.2 upgrade, so the new tests were added because of that.

### CODE REVIEW

**Comment by nyghtly-derek** [comment]
Thank you for working on this fix! Am I correct to assume that this will fix the same problems with keyword parameters as seen in https://github.com/aasm/aasm/pull/849?

Do you think this is ready for merge? Maybe we should ping a member of https://github.com/orgs/aasm/people?

**Comment by nyghtly-derek** [comment]
@anilmaurya Pinging you as the most recent maintainer. Do you think we're good to merge this fix for ruby 3.2?

File: spec/models/event_with_keyword_arguments.rb [hunk 1]
```diff
@@ -9,8 +9,16 @@ class EventWithKeywordArguments
...
     end
+
+    event :another_close do
+      before :_before_another_close
+      transitions from: :open, to: :closed
+    end
   end
 
   def _before_close(key:)
   end
```
Comment (msxavi) [inline spec/models/event_with_keyword_arguments.rb:20]: Can we please include a scenario where key is optional?
```
[1] pry(main)> def method_a(key:)
[1] pry(main)* end
=> :method_a
[2] pry(main)> def method_b(key: nil)
[2] pry(main)* end
=> :method_b
[4] pry(main)> method(:method_a).arity
=> 1
[5] pry(main)> method(:method_b).arity
=> -1
[6] pry(main)>
```

That should fall within `invoke_with_variable_arity` which from my understanding will fail.

**Comment by thiagofm** [comment]
@nyghtly-derek

I believe it is ready and fixed, I've ran it in production.

It could be that there's some unhandled / corner case, but in case that there is, we can fix as it shows up. Right now the test suite covers a lot of them.

I did send @anilmaurya an email a few months ago, he was busy at the time.

File: spec/models/event_with_keyword_arguments.rb (hunk 1 above)
Comment (thiagofm) [reply spec/models/event_with_keyword_arguments.rb:20]: My solution no longer relies on checking for arity, so it might still work.

I'll do this when I have time. If you want feel free to create a PR on my fork 💯

File: lib/aasm/core/invokers/literal_invoker.rb [hunk 2]
```diff
@@ -25,16 +25,57 @@ def subject_arity
...
+
+        def invoke_with_fixed_arity
           req_args = args[0..(subject_arity - 1)]
-          return record.__send__(subject, **req_args[0]) if req_args[0].is_a?(Hash)
-          record.__send__(subject, *req_args)
+          if req_args[0].is_a?(Hash)
+            record.__send__(subject, **req_args[0])
+          else
+            record.__send__(subject, *req_args)
+          end
```
Comment (derek-noredink) [inline lib/aasm/core/invokers/literal_invoker.rb:77]: Hey @thiagofm,

Thank you for working on this fix! I tried it out on my project and found what I think is a bug in your implementation. Here's the example case:

```
def sample_contents_present?(_client, sample)
  sample["content"].present?
end
```

And the relevant stack trace:
```
Error
wrong number of arguments (given 1, expected 2)
Stacktrace
./app/models/guided_draft/student_state.rb:328:in `sample_contents_present?'
.../lib/ruby/gems/3.2.0/bundler/gems/aasm-231f2fabedb4/lib/aasm/core/invokers/literal_invoker.rb:74:in `invoke_with_fixed_arity'
.../lib/ruby/gems/3.2.0/bundler/gems/aasm-231f2fabedb4/lib/aasm/core/invokers/literal_invoker.rb:48:in `invoke_with_arguments'
.../lib/ruby/gems/3.2.0/bundler/gems/aasm-231f2fabedb4/lib/aasm/core/invokers/literal_invoker.rb:31:in `exec_subject'
.../lib/ruby/gems/3.2.0/bundler/gems/aasm-231f2fabedb4/lib/aasm/core/invokers/literal_invoker.rb:19:in `invoke_subject'
.../lib/ruby/gems/3.2.0/bundler/gems/aasm-231f2fabedb4/lib/aasm/core/invokers/base_invoker.rb:48:in `invoke'
```

Our theory about why it's broken is essentially that we're missing a case for:
- fixed arity, with more than one argument, with the first argument being a hash

In this scenario, we do not want to only send the first argument, as you've done here:
```
if req_args[0].is_a?(Hash)
  record.__send__(subject, **req_args[0])
```

Rather, we want both arguments, that is:
```
record.__send__(subject, *req_args)
```

Does that make sense to you?

**Comment by alto** [comment]
@thiagofm Thanks for your contribution. This is a nasty bug you had been looking into here. ❤️ 

@thiagofm  @msxavi @derek-noredink Is there anything missing still? Otherwise I'm happy to merge in and release a new version for this.

**Comment by derek-noredink** [comment]
@alto Thank you for following up on this! We were actually able to upgrade to ruby 3.2 by using aasm 5.4. We only had to write a little work around for one of our tests.

In any case, I can't trace anything back to a problem with this pull request, so I'd say we're good to go. This is definitely an improvement!
